### PR TITLE
feat: oss storage support

### PR DESCRIPTION
**Pull request by masonyc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Adds ObjectStoreConfig::OSS for OSS object storage configurations.
- Create oss object-store instance in the datanode when configured.
- Introduced Config Structs for ObjectStoreConfig

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Added support for OSS storage according to this issue GreptimeTeam/greptimedb/issues/823

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
GreptimeTeam/greptimedb/issues/823

### CODE REVIEW

**Comment by masonyc** [comment]
Test Result on running `cargo test oss` :
image

image

**Comment by masonyc** [comment]
Test Result on running `cargo test oss` :
image

image

File: tests-integration/src/test_util.rs [hunk 1]
```diff
@@ -126,6 +166,7 @@ fn get_test_store_config(
 enum TempDirGuard {
     File(TempDir),
     S3(TempFolder),
+    Oss(TempFolder),
```
Comment (killme2008) [inline tests-integration/src/test_util.rs:169]: Prefer `OSS` which is the same as `ObjectStoreConfig`.

**Review by killme2008 — APPROVED** [review]
Great work, thanks a lot. There is a minor suggestion. LGTM.

File: tests-integration/src/test_util.rs (hunk 1 above)
Comment (waynexia) [reply tests-integration/src/test_util.rs:169]: `Oss` is fine according to [style-guide](https://pingcap.github.io/style-guide/rust/naming.html), we (me and @evenyag) had a discussion about this on https://github.com/GreptimeTeam/greptimedb/pull/850#discussion_r1064320339

Comment (killme2008) [reply tests-integration/src/test_util.rs:169]: It's okay, but it's better to keep it consistent between `TempDirGuard` and `ObjectStoreConfig`.

File: src/datanode/src/datanode.rs [hunk 2]
```diff
@@ -39,6 +39,13 @@ pub enum ObjectStoreConfig {
         endpoint: Option<String>,
         region: Option<String>,
     },
+    OSS {
+        bucket: String,
+        root: String,
+        access_key_id: String,
+        access_key_secret: String,
```
Comment (waynexia) [inline src/datanode/src/datanode.rs:46]: Do we need  to keep this field the same with S3's config above?

```suggestion
        secret_access_key: String,
```

File: src/datanode/src/instance.rs [hunk 3]
```diff
@@ -214,6 +216,36 @@ pub(crate) async fn new_object_store(store_config: &ObjectStoreConfig) -> Result
...
+    let (root, secret_key, key_id, bucket, endpoint) = match store_config {
+        ObjectStoreConfig::OSS {
+            bucket,
+            root,
+            access_key_id,
+            access_key_secret,
+            endpoint,
+        } => (root, access_key_secret, access_key_id, bucket, endpoint),
+        _ => unreachable!(),
+    };
```
Comment (waynexia) [inline src/datanode/src/instance.rs:229]: Maybe we can encapsulate a struct over these configs, like `struct OssConfig` and `struct S3Config`, and pass the corresponding struct to these constructor functions.

**Review by waynexia — COMMENTED** [review]
Thanks for your contribution @masonyc :heart:

File: src/datanode/src/instance.rs (hunk 3 above)
Comment (killme2008) [reply src/datanode/src/instance.rs:229]: Yep, good idea, refactor them into such as:

```rust
pub enum ObjectStoreConfig {
    File {
        data_dir: String,
    },
    S3(S3Config),
}

#[derive(Debug, Clone, Serialize, Default, Deserialize)]
#[serde(default)]
pub struct S3Config {
    pub bucket: String,
    pub root: String,
    pub access_key_id: String,
    pub secret_access_key: String,
    pub endpoint: Option<String>,
    pub region: Option<String>,
}
```

File: src/datanode/src/datanode.rs (hunk 2 above)
Comment (masonyc) [reply src/datanode/src/datanode.rs:46]: Hi @waynexia , in the opendal's implementation, the name for this field is not consistent with S3. I am happy to change it to `secret_access_key` if you want.

File: tests-integration/src/test_util.rs (hunk 1 above)
Comment (masonyc) [reply tests-integration/src/test_util.rs:169]: Renamed OSS to Oss for `ObjectStoreConfig`

File: src/datanode/src/instance.rs (hunk 3 above)
Comment (waynexia) [reply src/datanode/src/instance.rs]: `File` looks very outstanding, I prefer to wrap it as well 🥹

File: src/datanode/src/datanode.rs (hunk 2 above)
Comment (waynexia) [reply src/datanode/src/datanode.rs:46]: I did some search and it looks like this difference is coming from OSS and S3 official SDKs. Let's leave it unchanged...

File: src/datanode/src/instance.rs (hunk 3 above)
Comment (masonyc) [reply src/datanode/src/instance.rs]: Hi @waynexia @killme2008 , thanks for reviewing. I have updated the code to use the structs. Please have a review again when you have time.

**Review by waynexia — APPROVED** [review]
Looking good! Thanks again @masonyc

**Pull request merged**